Exposed toplevel status as a property for spatial and canvasitem

Duroxxigar opened the pull request:
Addresses #10453

Commit:
doc: Improve distance_fade documentation

Changed the documentation of distance_fade_max_distance and min
to be consistent with the behavior of the code as well as the user
documentation. Also noted the behavior of swapping max_distance and
min_distance (where max_distance < min_distance).

Fixes #36051.

Co-authored-by: Clay John <claynjohn@gmail.com>

Duroxxigar committed:
Exposed toplevel status as a property for spatial and canvasitem

Xrayez commented on doc/classes/CanvasItem.xml:
```diff
@@ -557,6 +557,9 @@
 			If [code]true[/code], the object draws behind its parent.
 		</member>
+		<member name="toplevel" type="bool" setter="set_as_toplevel" getter="is_set_as_toplevel">
+			If true, the node will not inherit its transform from parent [CanvasItem]s.
```
"true" should be wrapped as `[code]true[/code]` the way it's done for other descriptions for consistency, same for other changes.

Duroxxigar replied:
Fixed

kuruk-mm committed:
TileMap: Set texture_filter and texture_repeat to generated CanvasItems and update when it changes

timothyqiu committed:
Fixes crash when saving scene

aaronfranke committed:
Change Basis is_equal_approx to use instance method

aaronfranke committed:
Remove redundant is_equal_approx_ratio method

is_equal_approx is able to handle values of any size, and is_equal_approx_ratio is no longer used in any exposed APIs, so we don't need is_equal_approx_ratio anymore. Also, add #ifdef MATH_CHECKS for a method that is only used when MATH_CHECKS is defined.

bruvzg committed:
Adds PCK encryption support (using script encryption key for export).
Change default encryption mode from ECB to CFB.

Megamega53 committed:
doc: Some improvements to KinematicBody docs

Co-authored-by: Marcel Admiraal <madmiraal@users.noreply.github.com>
Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

follower committed:
Doc clarity/detail for extra spacing properties

Re: Space spacing being in addition to character spacing see:

 * <https://github.com/godotengine/godot/blob/d7b85fbaa1fc438effe406c9d7f973749eb4e527/scene/resources/dynamic_font.cpp#L858-L859>

Re: Value being able to be negative see example here:

 * <https://github.com/godotengine/godot/issues/38630>

But also note that nodes other than `Label` may not currently render extra space spacing correctly.

aaronfranke committed:
Rename the ".mono" folder to ".godot/mono"

Wavesonics committed:
HTTPRequest now accepts gzip
Added request_raw to HttpRequest
Added decompress_dynamic to Compression class
Added decompress_dynamic to BytePoolArray

Merge doc fix

revert

Wavesonics committed:
Fixed PackedByteArray GDScript functions

p_self._data.packed_array needs to be used instead of p_self._data._mem

printline

bruvzg committed:
Add window click-through support.

aaronfranke committed:
Rename the ".import" folder to ".godot/imported"

Arivval committed:
added load resource pack with offset feature

updated variables to use size_t

removed line break to make code style more consistent

added conditional check to return an error if offset field is used when loading a ZIP package

fixed typo

formatted file

added commit regarding self contained exe files

handled error loging for load zip file with offset

spelling tweak

updated conditional statement for magic check

udpated error message when load Zip file with offset is called

fix CI

Trying to fix CI

fix CI done

Added error message for loading self-contained exe with offset.

Updated documentation.

Fix indent

final fix indent

Updated documentation.

fix indents

Updated doc based on suggestion

Final fix

fixed format

Paulb23 committed:
Add and convert editor to use CodeEdit

Paulb23 committed:
Added gutter system to TextEdit

Paulb23 committed:
Added Line numbers to CodeEdit

Paulb23 committed:
Add main_gutter (breakpoints, bookmarks, execution lines) to code_edit

bruvzg committed:
[Complex Test Layouts] Change `String` to use UTF-32 encoding on all platforms.

Paulb23 committed:
Add folding gutter to code_edit

Paulb23 committed:
Changed line_edited_from(from) to lines_edit_from(from, to)

vnen committed:
Add GDScript code generation interface

Implement the abstraction by targeting the current VM.

Paulb23 committed:
Move ConnectionGutter to editor code_editor

Paulb23 committed:
Move safe line color into editor

Paulb23 committed:
Remove breakpoints from editor settings

Faless committed:
Better HiDPI support in HTML5.

Faless committed:
Make canvas resize optional in HTML5.

HaSa1002 committed:
Add C# code examples to the docs

Only existing GDScript code examples are converted and added to the
docs.
This is the first batch include classes beginning with A and B.

Included classes:
 * AcceptDialog
 * AESContext
 * Animation
 * AnimationNodeStateMachine
 * AnimationNodeStateMachinePlayback
 * AnimationNodeStateMachineTransition
 * Array
 * ArrayMesh
 * AStar
 * AStar2D
 * Bool
 * Button

fire committed:
Use the vertex colors by default in gltf.

Faless committed:
Expose request_quit method to JS in HTML5 export.

Faless committed:
Small refactor to JavaScript handlers.

Crated helper class in native/utils.js.
Simplify code in OS/DisplayServer.

Faless committed:
Window event listener do not use capture.

Faless committed:
[HTML5] Add override keyword, cleanup methods.

vnen committed:
Change GDScript compiler to use codegen abstraction

vnen committed:
Add GDScript disassembler

Commit:
Move GDScript tests to respective folder under modules

Commit:
Register GDScript test tools as test commands to run via command-line

reduz committed:
Re-Implement GPU particles on master.

-No new features yet
-Unlike godot 3.x, sorting happens using GPU

Lunatoid committed:
Fixes DefaultProjectIcon scaling with editor scale

Should fix #27009 where the DefaultProjectIcon was scaling
with the EDSCALE. Now it checks if the icon name is equal
to "DefaultProjectIcon" and sets the scale to 1.0 instead of
EDSCALE.

pouleyKetchoupp committed:
Fix drag and drop between windows in X11 display server

Proper implementation for get_window_at_screen_position:
Now getting the topmost last active window when overlapping.

Mouse drag & release events:
They are now propagated through the current focused window, in order to
make it consistent with the engine expectations and the Windows display
server implementation.

Calinou committed:
Document RigidBodies in character mode never sleeping automatically

See #7996.

Calinou committed:
Document how renaming the project affects the user data path

SecretPanda420 committed:
Remove GitHub actions badge.

`Godot` workflow is long removed in `master`, and is only found in the `3.2-android-app-bundle` branch currently. As of now Github actions has no way to show `passing` if all checks are passed but only to show the badge of each workflow individually. We can make a feature request later on but for now, we should remove the `Godot CI` badge as it reflects completely something else.

skyace65 committed:
Move note about using Resource in the File documentation

akien-mga committed:
Fix invalid CSV in SDL2 controller mappings DB

Follow-up to a61cae14696732fca214902161f40450dec2f9cd.

strank committed:
Fix doctest runner ignoring all arguments starting with --test

Instead of just the one argument that is exactly --test.
The long-form arguments --test-case and --test-suite were ignored.

Commit:
Fix PopupMenu unresponsive on right side of menu

EricEzaM committed:
Updated PopupMenu documentation after merge of #41640 - PopupMenu rework

EricEzaM committed:
Fix editor inspector refresh not working

kotelifelif committed:
Fix filtering of search results in SectionedInspector

madmiraal committed:
Merge python EnvironmentError, IOError and WindowsError into OSError.

ThakeeNathees committed:
fixed parser error when indexing a dictionary.

Fix: #41707

Yetizone committed:
Mention SubViewport.render_target_clear_mode property is intended for 2D usage.

vnen committed:
Add VariantInternal class

This is used to get a direct pointer for the data inside a Variant.
Added as another class instead of in Variant directly because this
should only be used in special cases, so use must be explicit.

Chaosus committed:
Fix particles shader to use built-ins

bruvzg committed:
Fix GDScript codegen leak.

Chaosus committed:
Restore Particles functionality in visual shader

neikeq committed:
C#: Fix Windows detection for copying MSBuild stub

Previous condition checked only the host
platform. This was a problem as our official
builds are from Linux.

madmiraal committed:
Ensure header guards enclose entire header.

EricEzaM committed:
Tooltip flickering and targeting fixes.

madmiraal committed:
Create a basic .lgtm.yml file.

neikeq committed:
Fix parsing of C# files with spaces in the path

neikeq committed:
Fix 'Parameter "assembly" is null' error

This error was normally being printed when
trying to open the project assembly while
the project was not yet built.
The error should not be printed. It's the job
of this method's caller to decide whether to
print an error or not if loading failed.

neikeq committed:
C#: Fix Godot failing to find class namespace

neikeq committed:
C#: Hide Build button if there's no solution to build

Same as we do with the bottom panel. Mainly to
avoid bothering if the project is not using C#.

akien-mga committed:
i18n: Sync translations with Weblate

(cherry picked from commit 41b9e1d7ea6241114babf0681c4e25400230ecfc)

madmiraal committed:
Ensure assignment operators return by reference to avoid unnecessary copies.

akien-mga committed:
doc: Sync classref with current source

ThakeeNathees committed:
null pointer dereference at GDScriptFunction::call fix

madmiraal committed:
Replace calls to gmtime with gmtime_r and localtime with localtime_r.

madmiraal committed:
Make PosVertex to Vertex slicing explicit.

lolleko committed:
Remove ArrayMesh::surface_remove declaration

Removal was omitted in 449df8f688080c02bfbbfafc45421875b77deb1b

bruvzg committed:
Fix Mono path utils char types.

Chaosus committed:
Improve messages in shader function validation

AnilBK committed:
Fixed Some typos.

Chaosus committed:
Fix crash regression in the shader function validation

introduced in #41780

Faless committed:
Update mbedTLS to version 2.16.8 (+ patch).

alexdlm committed:
Handle csproj "Remove" globs

MSBuild Item returns empty strings if an attribute isn't set (which
caused an IndexOutOfRangeException in NormalizePath).

We were treating Excludes incorrectly, Remove directives provide the
intended behaviour in the auto-including csproj format.

Chaosus committed:
Cleanup constructor code in visual shader nodes

Hassan-A committed:
Fix tilemap tiles on horizontal/vertical flips

EricEzaM committed:
Fixed get_usable_parent_rect() errors when initialising popup menus.

willnationsdev commented:
Since it was last updated in February, I'd recommend rebasing this pull request. Or, if unavailable, mentioning in the related Issue that someone else should handle the PR.

Chaosus committed:
Use flags instead TYPE_ enum in visual shaders editor

ThakeeNathees committed:
GDScript: parameter infer type bug fix

Fix: #41772

Chaosus committed:
Removes redundant code & fix documentation for VisualShaderNodeCustom

AndreaCatania committed:
Avoid adding margin twice along capsule Y axis

reduz committed:
Implement manual particle emission and particle sub emitters.

madmiraal committed:
Fix header guards in modules:
- Add missing header guards to various modules' register_types.h
- Add header guard to basis_universal/texture_basisu.h.
- Ensure header guard encloses entire header in
  webrtc/webrtc_data_channel_js.h.

Calinou committed:
Improve documentation related to Array error handling

This closes https://github.com/godotengine/godot-docs/issues/3834.

Calinou committed:
Clarify the required child node type in the GridContainer documentation

This closes #41822.

Chaosus committed:
Remakes particles in visual shaders

madmiraal committed:
Remove bit fields and use fixed width integers instead.

Chaosus committed:
Fix some broken visual shader nodes

madmiraal committed:
Use != 0 instead of > 0 when checking bit mask.

madmiraal committed:
Clarify comparison precedence when enabling or disabling back buffer.

EricEzaM committed:
Fixed expand/collapse folder option not showing in FileSystemDock

Faless committed:
Fix RasterizerDummy, override, cleanup.

Add override keyword to RasterizerDummy methods.
Plus cleanup unused methods, remove virtual keyword.

Chaosus committed:
Added `active` boolean to particles mode output in visual shaders

Razoric480 committed:
Add a VScript func to sanitize variant values

Fixes #27611

pouleyKetchoupp committed:
Fix LineEdit not consuming enter events

LineEdit should not return early when processing KEY_ENTER, so it can
consume the event properly.

Regression introduced by mistake while fixing enter events for Android
(PR #40487 - c0b394572f35498801571ad7176eb357d5de1bf3)

pouleyKetchoupp committed:
Fix TURN server example in WebRTC documentation

WebRTC GDNative plugin uses `credential` and not `credentials`.
https://github.com/godotengine/webrtc-native/blob/74f2c78db5cdffa5b2b6ba9cd041061d7694400c/src/WebRTCLibPeerConnection.cpp#L35-L37

madmiraal committed:
Check if old mouse column is still available.

madmiraal committed:
Initialise Basis elements with a default Basis in the declaration.
Ensures a valid Basis is created with all constructors.

akien-mga committed:
i18n: Sync translations with Weblate

Add translators comment for the two 'Project Manager' strings,
and fix some of the translations.

(cherry picked from commit c5f6d2097bcbe1826b9111774ed0f8deb5839060)

Yetizone committed:
mesh_library.h: Update header guard to reflect file name

akien-mga committed:
Update AUTHORS and DONORS list

New contributor added to AUTHORS:
@EricEzaM

Thanks to all contributors and donors for making Godot possible!

neikeq committed:
C#: Fix endless reload loop if project has unicode chars

The assembly modified time wasn't picked properly
as the path was treated as latin-1, so the file
watcher was constantly firing the event.

neikeq committed:
C#: Use BOM when creating a solution

At least on Windows there seems to be issues if
the solution has no BOM and contains a project
with cyrillic chars.

YeldhamDev committed:
Add 'get_previous_tab()' to 'Tabs'

aaronfranke committed:
Change inequality comparison operators to use exact equality

11thPenguin committed:
Fixed minor typo, gui_hid -> gui_hide.

Chaosus committed:
Improve performance of Undo:change node position in visual shader

pouleyKetchoupp committed:
Fix general keyboard input lag on X11 display server

This change makes keyboard inputs more responsive on Linux, especially
when the FPS is lower on slower configurations.

Polling events from the x server is done on a separate thread to avoid a
frame delay with inputs, due to first sending the event to the input
manager with XFilterEvent then processing the new event only on the next
frame.

Calls to Input Manager functions like XSetICFocus, XUnsetICFocus and
XSetICValues use a mutex, because they are polling events internally and
would otherwise interfere with our own thread process for polling events
which can cause a deadlock in some cases.

XUnsetICFocus is called instead of XSetICFocus on FocusOut events,
so the input manager can be properly notified of focus changes.

clipboard_get now uses a blocking call to poll for a specific event type
when waiting for a SelectionNotify event, instead of polling all events
and filtering them afterwards.

akien-mga committed:
GDScript: Fix MutexLock usage, fixes Clang 6 compat

Fixes #41852.

madmiraal committed:
Explicitly add implicitly added semicolons.

skyace65 committed:
Document some properties of NavigationMesh

Chaosus committed:
Improve performance for Show/Hide port preview in visual shaders

reduz committed:
Implement 3D textures as import and resource format.

Chaosus committed:
Fix ColorPickerButton to apply changes after picker popup closed

akien-mga committed:
stb_vorbis: Increase max alloc buffer size for big Vorbis comments

The previous max worked OK for audio data, but stb_vorbis recently gained support
for Vorbis comments, which can embed up to 2^32-1 bytes of data (e.g. cover art
encoded as base64).

We use 2^30 as max which should be sufficient for most files.

Fixes #41913.

Chaosus committed:
Fix build warning in visual_shader_editor

reduz committed:
Make EXR import format support all depths and channel configurations

Calinou committed:
Rename ShortCut to Shortcut which is more grammatically correct

See https://github.com/godotengine/godot/issues/16863#issuecomment-685236980.

akien-mga committed:
tinyexr: Sync with upstream 1.0.0

RandomShaper committed:
Prevent cyclic reference between script and its members

RandomShaper committed:
Ensure cyclic dependencies between scripts are broken at exit

clayjohn committed:
Set repeat in all 3 dimensions in 3d textures

timothyqiu committed:
Fixes ParallaxLayer offset when ignore camera zoom

Chaosus committed:
Added Texture3D to visual shaders

Chaosus committed:
Improve performance of Add/Remove/Connect/Change nodes in visual shader

madmiraal committed:
Remove unused variable in fixed-size.html.

madmiraal committed:
Remove duplicate or unnecessary Python definitions.

madmiraal committed:
Remove unused Python imports.

capnm committed:
git: Ignore clangd cache folder

Yetizone committed:
scene_tree.h: Update header guard

Calinou committed:
Rename RayCast's `cast_to` property to `target_position`

`cast_to` is sometimes mistaken as a method rather than a property.
`target_position` makes it more obvious that it's a property.

skyace65 committed:
Improve SpriteFrames get_animation_loop description

Calinou committed:
Update and complete the Sprite3D and SpriteBase3D class documentations

This removes an outdated notice about Sprite3D performance issues
which will most likely be fixed in the new implementation.

lyuma committed:
Fix ATTENUATION and implement SHADOW_ATTENUATION.

pouleyKetchoupp committed:
Fix RichTextLabel alignment for clickable regions

Fixes #41006 (regression from #39164).

The original alignment fix was limited to PROCESS_DRAW mode, which
caused some discrepancies with PROCESS_POINTER mode.
Now only PROCESS_CACHE is excluded with a condition a few lines above.

akien-mga committed:
doc: Sync classref with current source

Bind missing enums.

madmiraal committed:
Remove unused Python local variables.

madmiraal committed:
Add missing reference to const T Vector::get().

Xrayez committed:
Move GDNative `String` tests to respective module

GDNative-specific tests moved out of main `tests/` folder into
`modules/gdnative/tests`.

Include path for GDNative headers are still hardcoded in `tests/SCsub`,
but made conditional now.

Also fixed test case tag typos.

Co-authored-by: bruvzg <7645683+bruvzg@users.noreply.github.com>

Chaosus committed:
Fix triplanar texture code generation in visual shaders

ebrahimebrahim committed:
Clarify get_data Image method

skyace65 committed:
Improve AudioEffectDistortion and AudioEffectFilter documentation

Rubonnek committed:
Disconnect item_rect_changed when removing a child of GraphEdit

Chaosus committed:
Update shader graph if mode is changed

HaSa1002 committed:
Translate GDScript Code Examples to C# (C)

Translates Code Examples in classes beginning with `C`.

Includes:
 * Callable
 * CanvasItem
 * CharFXTransform
 * Color
 * ColorRect
 * ConfigFile
 * ConfirmationDialog
 * Control
 * Crypto

vnen committed:
Fix Vector{2,3}i casting on Variant multiplication with numbers

Xrayez committed:
Fix `List` swap behavior on front, back and adjacent elements

- immediately return if A == B;
- first and last elements (front, back) are updated upon relinking;
- handles a special case of forward and backward adjacent elements.

Xrayez committed:
Add a test suite for `List`

Adds a number of test cases for `List::swap()`.

aaronfranke committed:
Add concatenation support and a new ctor to Godot.Collections.Array

Commit:
striped edges on node rename in SceneTreeDock

neikeq committed:
C#: Fix csproj not synced on file move/removal from FS dock

When NormalizePath was called with an absolute
path (with drive letter) on Windows, it would
prepend a file path separator to the path, e.g.:
'\C:\Program Files\'.
Apparently this was still accepted as a valid
path by DotNetGlob and it stopped working when
we switched to MSBuildGlob.

Commit:
creater user-dir, if non-existant and pressing 'Open Project Data Folder'

tdaffin committed:
Add mono log message to error for fatal errors

JFonS committed:
Improve rotation gizmo

Hide the back sides of the rotation gizmo circles and add a white
outline for better visualization of the rotation "sphere".

Chaosus committed:
[VisualShader] Fixes default node's input port not hiding on connection

Calinou committed:
Document how to remove metadata from an Object

This closes https://github.com/godotengine/godot-docs/issues/4010.

Chaosus committed:
Fix visual shader connection regression

hoontee committed:
Fix "Create Trimesh Collision Sibling" transform

dalexeev committed:
Several documentation improvements

madmiraal committed:
Fix RigidBodyBullet areasWhereIam element shift.

Calinou committed:
Only display the Windows toggle console option if it can actually be used

dalexeev committed:
Improvement for the Copy button in the Output Log

Now if no text is selected, pressing the Copy button copies the entire text.

Scony committed:
Add test cases for numeric literals with underscores

Razoric480 committed:
Make LSP ignore $/ messages

Fixes #38814

Commit:
Dropping file on parrent dirrectory fixed

Chaosus committed:
Renames Type to OpType in VisualShaderNodeMultiplyAdd

To prevent possible conflicts with C# and other languages.

akien-mga committed:
makerst: Add missing newline at EOF

melvin-ibuildings committed:
Show correct name of signal when editing an existing one

Use popup_dialog() instead of popup_centered() to show edit dialog/

With popup_centered it is not possible to change the displayed
signal name. When this is not set the previous shown name is show
for the current dialog.

This is no problem when creating a new conenction as popop_dialog
is used there and this would update the title.

Fixes #42074

Calinou committed:
Clarify that 2D cell sizes are defined in pixels

See https://github.com/godotengine/godot/issues/40577#issuecomment-662329021.

skyace65 committed:
Add documentation for shelf filters

Calinou committed:
Mention compatibility-breaking Camera2D offset change in the changelog

See https://github.com/godotengine/godot/issues/42088.

Xrayez committed:
Make `Object::to_string` virtual

Allows to override printing via C++, not only via script.

Faless committed:
Fix certificate generation with mbedtls 2.16.8 .

When generating certificates with
`Crypto.generate_self_signed_certificate` we generate the PEM in a
buffer via `mbedtls_x509write_crt_pem`.

Since version 2.16.8, mbedtls adds spurious data at the end of the
buffer due to internal optimizations, this breaks our logic when we try
to immediately parse it and return a proper `X509Certificate` object.

This commit updates the code to find the actual PEM length to parse
using `strlen`, takes extra caution always adding the terminator to the
buffer, and slightly improve error messages.

raulsntos committed:
Fix ExprMatch stackoverflow

Calinou committed:
Cross-reference AABB getter methods in the documentation

This also clarifies that `get_aabb()` returns the AABB in local space.

See https://github.com/godotengine/godot/issues/42095.

akien-mga committed:
Donors: Add Gamblify, remove Interblock from splash screen

DrMoriarty committed:
iOS: Fix multiple issues with PVRTC import, disable ETC1

Fixes: #28683, #28621, #28596 and maybe others

For iOS we enable pvrtc feature by default for all backends
Etc1 for iOS doesn't have any sense, so it disabled.
Fixed checks in export editor.
Fixed pvrtc encoding procedure.

Edit by Akien: Forward-ported from #38076, this may not make sense as is for
Godot 4.0, but it's important that we have the latest code in sync with 3.2
for when more rendering backends and proper iOS support are added back.

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

andriyDev committed:
Added inspector_only option to inspect_object in EditorInterface.
Updated docs to reflect change to inspect_object.

Calinou committed:
Improve documentation about VideoPlayer video formats

This closes https://github.com/godotengine/godot-docs/issues/4021.

kotelifelif committed:
Fix changing icons' colors when switch dark and light theme

clayjohn committed:
expose local RenderingDevice creation to RenderingServer

vnen committed:
Allow commutative multiplication in Variant

Also allow quaternions to be multiplied by integers.

Vinealius committed:
Fixes the resize of tile shapes when a vertex is outside of the tilesheet. Fix #34970

akien-mga committed:
Linux/BSD: Fix support for NetBSD

Add __NetBSD__ to `platform_config.h` so that it can find `alloca`
and use the proper `pthread_setname_np` format.

Rename RANDOM_MAX to avoid conflict with NetBSD stdlib.

Fixes #42145.

akien-mga committed:
NetBSD: Implement OS_Unix::get_executable_path()

Same implementation as OpenBSD seems to work fine.

akien-mga committed:
X11: Try to load libXrandr.so.3 if libXrandr.so.2 isn't found

All Linux distros, and FreeBSD and OpenBSD seem to have libXrandr.so.2,
but for some reason recent NetBSD versions seem to have libXrandr.so.3 now.

akien-mga committed:
Fix typos with codespell

Using codespell 1.17.1.

Method:
```
$ cat > ../godot-word-whitelist.txt << EOF
ang
curvelinear
dof
doubleclick
fave
findn
leapyear
lod
merchantibility
nd
numer
ois
ony
que
seeked
synching
te
uint
unselect
webp
EOF

$ codespell -w -q 3 -I ../godot-word-whitelist.txt --skip="./thirdparty,*.po"
$ git diff // undo unwanted changes
```

Faless committed:
JS synchronous start, better persistent FS sync.

The engine now expects to emscripten FS to be setup and sync-ed before
main is called. This is exposed via `Module["initFS"]` which also allows
to setup multiple persistence paths (internal use only for now).

Additionally, FS syncing is done **once** for every loop if at least one
file in a persistent path was open for writing and closed, and if the FS
is not syncing already.

This should potentially fix issues reported by users where "autosave"
would not work on the web (never calling `syncfs` because of too many
writes).

Listwon committed:
Fix custom modules in Visual Studio

pouleyKetchoupp committed:
Fix popup menu item selected when opening the menu

In order to allow selecting items by either holding left click, or click
to open and click again to select, mouse button release was invalidated
based on the amount of mouse motion.

This was causing issues in some scenarios where an item could be
selected while opening the menu if the mouse moved enough between button
press and release.

This case could happen in the language selection of the project manager,
especially on linux, because of the order and timing of the mouse
events on x11.

This change invalidates mouse release based on a timing condition rather
than moved distance to handle any case from the display server properly.

akien-mga committed:
zstd: Update to upstream version 1.4.5

Commit:
add description for ScriptEditor.open_script_create_dialog()

m4gr3d committed:
Add overridable init method for the Godot fragment instance.

Commit:
add iOS Simulator platform 

without this we have to manually drop a `x86_64` builds of gdnative libs into the Xcode project to allow running in `iOS Simulator`

Calinou committed:
Fix crash when opening a ZIP data pack

Co-authored-by: bruvzg <7645683+bruvzg@users.noreply.github.com>

Listwon committed:
Add all headers to VS Project

rcorre committed:
Fix hints on PhysicsMaterial bounce/friction.

These values are only meaningful in the range 0 to 1.
Make sure the editor enforces reasonable values.
Fixes #42202.

Calinou committed:
Document that the JavaScript singleton may be disabled at build-time

Chaosus committed:
Forbids varying declaration in particles and sky shaders

Calinou committed:
Document `RichTextLabel.append_bbcode()` not closing previously open tags

This closes #42216.

Chaosus committed:
Fix some bugs in visual shader editor

Calinou committed:
Cross-reference AABB and Rect2 in the class reference

Calinou committed:
Fix a typo in the Array class documentation

This closes https://github.com/godotengine/godot-docs/issues/4049.

Chaosus committed:
Fix UniformRef invalid updating when LineEdit focus out

Chaosus committed:
Fix expressions nodes in visual shaders

zaevi committed:
Fix C# string.IsAbsPath()

kotelifelif committed:
Fixes updating CPUParticles emmision shape values

Faless committed:
Ignore thread models when compiling with NO_THREAD

The thread model option for physics (2D) and rendering (single-unsafe,
single-safe, multithread), was causing crashes/locks when set as
multithreaded and exported for a platform that does not support threads
(namely HTML5).

This commit ensures that when threads support is not available, that
option is ignored, and the equivalent of "single-unsafe" is always used
instead.

KoBeWi committed:
Check for null in InputMap::_find_event

Calinou committed:
Improve documentation related to printing error/warning messages

Calinou committed:
Reference the BaseButton class explicitly in button class documentations

Users are sometimes confused as to the `pressed` property not being
visible in the Button class documentation. This is because `pressed`
is defined in BaseButton.

Calinou committed:
Reference the Color constants cheatsheet in the class reference

ricardoalcantara committed:
Basis RotationQuat should be public.

akien-mga committed:
MessageQueue: Change default queue size to 4096 KB

1024 KB was low enough that many users seem to hit it, which can lead to the
editor freezing.

The proper fixed as described in #35653 would be to implement a page allocator
to prevent this overflow, but as a stop-gap measure, we can increase the
default value to a more lenient 4096 KB which should be high enough for the
vast majority of use cases.

The default size can be brought down again if/when #35653 is properly fixed,
and if it's actually relevant from a memory point of view.

KoBeWi committed:
Remove unnecessary empty line from Dictionaries

akien-mga committed:
i18n: Sync translations with Weblate

(cherry picked from commit 510d87e468d19836c9946830ffd3a6edcfaebf04)

akien-mga committed:
OS: Remove unused get_splash_tick_msec

It was added in 3e20391bf607dc7c452b056854aed4a8c99ba0f6 but it doesn't seem
particularly useful, and it was only implemented for the custom splash branch
and not the default one, so it could return an uninitialized int.

akien-mga committed:
i18n: Rename pt_PT to pt

We originally used `pt_PT` (i.e. Portuguese (Portugal)) to distinguish with
the Brazilian Portuguese variant `pt_BR`, as both are significantly different
and need separate translation files.

But Portugal's Portuguese (or "European Portuguese") is close to the variant
spoken and written in other Portuguese-speaking countries such as Angola and
Mozambique, so it makes sense for users of these countries to also have access
to the European Portuguese translation (at least until translators decide that
adding e.g. `pt_AO` and `pt_MZ` variants would make sense, taking into account
the translation effort that this duplication implies).

Godot's locale matching checks first for the full locale (e.g. `pt_AO`), and
if no translation is found, it checks for the non-regional language code
(`pt`), so this change enables translations for Portuguese speakers outside
Portugal and Brazil.

Calinou committed:
Add a property hint to StyleBoxFlat `shadow_size` for editor usability

This adds a visible range to the slider so it can be dragged more easily.

This closes #42309.

timothyqiu committed:
Surrounds boolean false with code tag

bruvzg committed:
[macOS] Fix mouse position in captured mode.

Calinou committed:
Enable the `copy_mono_root` SCons option by default

This closes #41652.

Calinou committed:
Replace AutoKey icon with a more universally understood "record" button

See discussion in
https://github.com/godotengine/godot-proposals/issues/169.

akien-mga committed:
Revert "Fix editor inspector refresh not working"

akien-mga committed:
Style: Fix black formatting after #42332

Somehow it did not run CI checks so we missed that one.

Also pin `black` version to latest upstream release.

pouleyKetchoupp committed:
Fix delay to process clipboard content from Godot in other programs

When pasting clipboard content from Godot to other applications,
multiple SelectionRequest events are sent to Godot in order to access
the data. It could take a long time before the data is ready for the
other app because events were processed one by one on the main thread,
especially when Godot is unfocused and runs at low frequency.

With this change, SelectionRequest events are directly handled on the
separate event polling thread to minimize this delay.

This change also replaces clipboard_get() calls in SelectionRequest with
a direct access to internal_clipboard, since in this case we know Godot
is the owner of the clipboard content and it's not necessary to query
the x server for it.

Rubonnek committed:
Allow retrieval of FuncRef function string

RandomShaper committed:
Avoid infinite loop in GDScript at shutdown

Chaosus committed:
Fix undo for moving multiple visual shader nodes

Rubonnek committed:
Make "function" a property of FuncRef

groud committed:
Fix error when dragging anchors with parent's size == 0

NathanLovato committed:
Remove link to step-by-step/animations

Required for godotengine/godot-docs#4074 to pass, we're removing that
page from the docs as part of the getting started section rewrite.

groud committed:
Draw selected item on top of everything else

Pennycook committed:
Make YSort stable

Keeps track of the order in which items are collected by
_collect_ysort_children, and uses that order to break
ties between items with similar Y positions.

Calinou committed:
Allow using a comma as decimal separator in EditorSpinSlider

This closes https://github.com/godotengine/godot-proposals/issues/1576.

timothyqiu committed:
Hides special folders in FileDialog for macOS

Craig-Stoneham committed:
Removed unnecessary keyword, fix comment style

The privacy here is already private

KoBeWi committed:
Don't warp mouse when selecting tracks

Calinou committed:
Automatically start searching in the asset library when entering text

Entering text will now start searching automatically after 0.25 seconds
have passed (debounce delay).

This removes the need for a separate Search button.

dalexeev committed:
Improve appearance of [connection] and [editable] sections in .tscn files

Chaosus committed:
Drag & drop 3d/2d array/cubemap texture to visual shader

eduardohmrodrigues committed:
Fix error message when exporting a write-only property without a setter

Calinou committed:
Make the pressed autokey icon red to emphasize its "recording" status

timothyqiu committed:
Fixes FileSystem tree preview icon size on HiDPI

Calinou committed:
Improve the String comparison methods' documentation

This closes https://github.com/godotengine/godot-docs/issues/4096.

akien-mga commented:
We actually discussed this in a live Q&A today, exposing this as a property makes sense. It needs a rebase before it can be merged.

Might be worth mentioning in the docs how child nodes of a "toplevel" node behave if they are not toplevel themselves (see #24154).

Duroxxigar commented:
uhhh - oops. This is why you don't work on stuff so late in the night I guess. Really don't feel like fixing this PR.

Closing and just created a new one - #42448
